### PR TITLE
Add support for using a different template for combox closed state

### PR DESCRIPTION
**Pull request by almightyju**
## What does the pull request do?
It provides a way to use a different template for items in the dropdown of a combo box separate to the closed state. If `SelectedItemTemplate` is not set the existing functionality is retained by using `ItemTemplate`

## Fixed issues
Addresses point 1 of https://github.com/AvaloniaUI/Avalonia/issues/7462

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=9689) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=9689) before we can accept your contribution.<br/><br/>:x: almightyju [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=9689)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=9689) it.</sub>

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027511-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by Arhell — COMMENTED** [review]
Please sign CLA

File: src/Avalonia.Controls/ComboBox.cs [hunk 1]
```diff
@@ -173,6 +188,24 @@ public VerticalAlignment VerticalContentAlignment
             set { SetValue(VerticalContentAlignmentProperty, value); }
         }
 
+        /// <summary>
+        /// Gets or sets the data template used to display the item in the combo box (not the dropdown)
+        /// </summary>
+        public IDataTemplate? SelectedItemTemplate
```
Comment (grokys) [inline src/Avalonia.Controls/ComboBox.cs:194]: What does this property do? I can't see it used anywhere. What's the difference between this and `DisplaySelectedItemTemplate`?

**Comment by grokys** [comment]
Is this the same as [`SelectionBoxItemTemplate`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.combobox.selectionboxitemtemplate?view=windowsdesktop-7.0) in WPF or does it differ? Should we be using the WPF naming?

**Comment by almightyju** [comment]
> Is this the same as [`SelectionBoxItemTemplate`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.combobox.selectionboxitemtemplate?view=windowsdesktop-7.0) in WPF or does it differ? Should we be using the WPF naming?

I've actually never come across that so I'm not familiar with it but it's read only anyway so I'm not sure using the same name would be the best idea since it wouldn't be usable?

The way I had this implemented in WPF was using an [`ItemTemplateSelector`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.itemscontrol.itemtemplateselector?view=windowsdesktop-7.0), much like this [stack overflow post](https://stackoverflow.com/a/36017186)

**Comment by timunie** [comment]
I think our equivalent for `ItemTemplateSelector` is implementing `IDataTemplate` or `FuncDataTemplate`.  Don't know if that'd help you anyway.

**Comment by almightyju** [comment]
> I think our equivalent for `ItemTemplateSelector` is implementing `IDataTemplate` or `FuncDataTemplate`. Don't know if that'd help you anyway.

I'm not sure about `FuncDataTemplate` since I cant see how to use it but it appears to have the same limitation as `IDataTemplate` where you can't control the template based on where it is within the UI.

In theory it's possible to have a bool in a viewmodel of `DropDownOpen` and use that to select a different template however I think that would also change the closed state (which is also visible when the dropdown is open) and polluting the viewmodel with view state feels a bit hacky.

If the goal is to be 100% WPF compatible `ItemTemplateSelector` is the way to go and I did have a look at how to go about implementing that but the `ItemContainerGenerator` stuff seemed a bit beyond my use case. 

I think having the ability to easily set a different template for a closed state for a combo box is better than the complications of using a template selector; for a Listbox/ItemsRepeater it makes sense to use `IDataTemplate` since you're binding to view model items anyway

**Comment by timunie** [comment]
> > I think our equivalent for `ItemTemplateSelector` is implementing `IDataTemplate` or `FuncDataTemplate`. Don't know if that'd help you anyway.
> 
> I'm not sure about `FuncDataTemplate` since I cant see how to use it but it appears to have the same limitation as `IDataTemplate` where you can't control the template based on where it is within the UI.
> 
> In theory it's possible to have a bool in a viewmodel of `DropDownOpen` and use that to select a different template however I think that would also change the closed state (which is also visible when the dropdown is open) and polluting the viewmodel with view state feels a bit hacky.
> 
> If the goal is to be 100% WPF compatible `ItemTemplateSelector` is the way to go and I did have a look at how to go about implementing that but the `ItemContainerGenerator` stuff seemed a bit beyond my use case.
> 
> I think having the ability to easily set a different template for a closed state for a combo box is better than the complications of using a template selector; for a Listbox/ItemsRepeater it makes sense to use `IDataTemplate` since you're binding to view model items anyway

I agree 👍. A different DataTemplate would be the easiest and most flexible solution.

**Comment by almightyju** [comment]
I thought of a better way to do this since I agree having two properties was a bit off. I hadn't put that much thought into it.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027866-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by amwx** [comment]
A couple of (late) thoughts here:
As pointed out above, WPF solves this via ItemTemplateSelector, which Avalonia doesn't use in favor of the implicit data template approach, which is one of the few limitations of this way.
Just thinking real quickly, Avalonia's styling system is more powerful and ContentTemplate is a StyledProperty, so you might be able to declare the template as a resource and use a style selector to override the template property on the ContentPresenter:

```Xaml
<Style Selector="ComboBox.diffTemplate /template/ ContentControl#ContentPresenter">
    <Style Selector="ContentTemplate" Value="{StaticResource MySelectionBoxTemplate}" />
</Style>
```

Actually, just did a quick test, and it seems this doesn't work since `TemplatedParent` priority from the template is winning over the `Style` priority of the custom style. I wonder if those priorities should be switched? However, you could get a local value priority if you dropped into code-behind/wrote a custom behavior to dig out the selection box ContentControl. I'm not sure how common of a use case this is that warrants this in the main ComboBox implementation.

**Comment by amwx** [comment]
Re today's update:
Not 100% sure how I feel about the PriorityValue approach added and whether it'd just be better to just set the template from the code of ComboBox when the content is set. Even now, the SelectionBox doesn't support the implicit DataTemplate approach (not specifying `ItemTemplate`), and this approach would make that harder to fix. But I'm not an expert here so I'll digress on this.

As for the editable combobox, that needs to be in its own PR as its unrelated to the original purpose of this PR. In addition, making ComboBox editable needs a bit of an API review beforehand. There's a lot to think about for this (I have an outdated/deprecated attempt at this following UWP's ComboBox that I'm trying to redo), two of the bigger items being:
- `SelectedItem`/`SelectedIndex` don't like it when its set to an item not present in `Items`, which at least from UWP, setting the Text property to something custom sets that as the SelectedItem even its is not in the Items list.
- Text input/selection/searching needs to respect DisplayMemberBinding (DisplayMemberPath in WPF/UWP) and not just rely on object.ToString(). WPF's method uses reflection and Avalonia is trying to move away from that and doing this in a reflection free way is quite complicated. (Similarly if something equivalent to `SelectedValue`/`SelectedValuePath` is added)

**Comment by robloo** [comment]
> As pointed out above, WPF solves this via ItemTemplateSelector

Good point, perhaps this is part of a more general case. Separately, I also have the need to select one DataTemplate when an item is selected and another DataTemplate for any other items that are not. There is no great way to do this that I've found yet and all options I've tried (1. custom IDataTemplate implementation and 2. swapping DataTemplate in Styles based on selection) don't work quite right in Avalonia (there are bugs).

But this same problem seems to apply here as well. If Avalonia had a good way (that worked with virtualization, etc.) of switching the DataTemplate based on whether an item is selected or not this could, unless I'm missing something, work here as well. Whatever is in the SelectedItem would have the selected DataTemplate applied and all else would have the default non-selected DataTemplate. Of course the items in the drop-down would all have to be considered "not selected" working values.

Anyway, just a thought based on other things that have come up recently.

**Comment by almightyju** [comment]
> ```
> <Style Selector="ComboBox.diffTemplate /template/ ContentControl#ContentPresenter">
>     <Style Selector="ContentTemplate" Value="{StaticResource MySelectionBoxTemplate}" />
> </Style>
> ```

I tried this myself while trying to find a workaround and also got nowhere, and even if it did work it feels really unintuitive since devs would need to look up the internals of the combobox to see how the template is being used, compared to intellisense showing `SelectedItemTemplate` when typing `ComboBox.ItemTemplate`.

> you could get a local value priority if you dropped into code-behind/wrote a custom behavior to dig out the selection box ContentControl. I'm not sure how common of a use case this is that warrants this in the main ComboBox implementation.

I think an answer of "learn the framework and do this rather complex thing" is a bit off regardless of known use cases (given that the first few results on google of doing this is from 2011 I'd say it's not that uncommon). Coming from WPF there's no easy way to achieve this outside of writing your own control or possibly behaviour which I think shouldn't be the case. 

> Not 100% sure how I feel about the PriorityValue approach

This came about because I found a few uses of it in the code I'm trying to port and figured since it's a WPF thing it wouldn't be a big deal to introduce it. I don't see a functional way to achieve the same result without some form of converter or custom binding.

> As for the editable combobox ...

This was a bit of an oversight since I'm using this branch locally and pushed it without thinking of this PR. Easy enough to revert but the main topic seems to be blocked by design decisions rather than implementation details so I'll stick to my branch for now since it does what I need.

**Comment by jmacato** [comment]
Closing this PR temporarily due to inactivity. Please ping me if this needs to be reopened.

**Pull request closed without merging**